sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyService.java

```diff
@@ -115,6 +115,7 @@
 
     @Get("keys/{key-name}/versions")
     @ExpectedResponses({200})
+    @UnexpectedResponseExceptionType(code = {404}, value = ResourceNotFoundException.class)
```

Comment:
This doesn't abide guidelines: https://azure.github.io/azure-sdk/general_implementation.html#error-handling

1. The server error is not exposed, which is reuqired.
2. We shouldn't create derived types when the user has mitigation, like they can check for 404 themselves, which is what we do in other languages.

I think the documentation should be fixed instead of wrapping the service exception (which you're not currently doing anyway) in another one.

Reply:
You can check my interpretation with architects.

Reply:
Thanks for the feedback Heath! About the points you raise:

1. We do include the server error in both [HttpResponseException](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/exception/HttpResponseException.java) and [ResourceNotFoundException](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/exception/ResourceNotFoundException.java).

2. Both exceptions mentioned above are a part of azure-core for Java, so we would not be creating a new derived type, just using a different one that is more specific to the case. In any case, since there are not 

I do agree the documentation should be fixed since it doesn't reflect current behavior.

One thing I'm not sure about, is the use of exceptions for server errors that are not specified in a service's swagger: in this case, it specifies that a successful response will only have a status code `200`. Since the guidelines say that we should produce an error for all other cases, it seems ok to me to produce an exception like the ones I mentioned, we even [do it](https://github.com/Azure/azure-sdk-for-java/blob/5535278916cc52dec876185a37421e2cf66a0246/sdk/keyvault/azure-security-keyvault-keys/src/main/java/com/azure/security/keyvault/keys/KeyService.java#L130) for endpoints like `/keys/{key-name}/backup` where [there is no specific mention](https://github.com/Azure/azure-rest-api-specs/blob/c9084e3a184c938d4110345293b410b64fc4bd65/specification/keyvault/data-plane/Microsoft.KeyVault/stable/7.0/keyvault.json#L378) of a `404` status code.

I will also be bringing this up with an architect like you suggested :)

Reply:
as per java guidelines, we should throw ResourceNotFoundException when service sends back 404.

Reply:
Should we add this same annotation to the other endpoints that don't have it? Isn't this considered a breaking change?

Reply:
All these would be breaking changes, I assume, including this PR.

Reply:
Yes, this would be a breaking change. To note, we don't throw an exception for .NET, and I don't believe Python or TS do as well. I think the code needs to be left as-is and the documentation fixed to reflect reality.

Reply:
This would not be a breaking change, I believe.
We currently throw HttpResponseException.
This change will throw ResourceNotFoundException which extends HttpResponseException.

Reply:
Except the discussion in the issue was about not returning a 404, but instead a 200 with an empty list. In what cases are you getting a 404 where you didn't during your investigation? If an empty list was returned before, and not you throw an exception - any exception - that is a breaking change.

Reply:
For example, .NET doesn't throw - it just gets an empty enumerable. Is that desirable? Probably not, but it's what the server returned to us. So I'm not even sure this fix would do anything since the service is not returning a 404 (proved with Postman earlier as well).

Reply:
JS and Python also return an empty list / don't throw.

Reply:
So, my understanding of all this is:

- Don't throw when we get a server response with an empty list and status code `200`
- It is ok to throw a more specific exception that extends from the one we already throw today.

Is this correct?

Reply:
If the response is 200, we don't throw any exceptions currently and return the empty list already.
When we get 404, as the user reported issue #11033  mentions, we should throw the more specific exception.

Reply:
*If* they get a 404, yes, but @vcolin7 was unable to reproduce that. That said, I guess I don't see a harm in this as long as you're sure you're not throwing when an empty result is returned (i.e. 200).